src/pagination/index.tsx

```diff
@@ -0,0 +1,178 @@
+	initialPage?: number;
+	initialPageSize?: number;
+	currentPage: number;
+	pageSize: number;
+}
+
+const icache = createICacheMiddleware<PaginationCache>();
+const factory = create({ theme, icache, i18n }).properties<PaginationProperties>();
+
+export default factory(function Pagination({ middleware: { theme, icache, i18n }, properties }) {
+	const {
+		initialPage,
+		initialPageSize,
+		onPageChange,
+		onPageSizeChange,
+		pageSizes,
+		pageSizeRenderer,
+		siblingCount = 3,
+		total
+	} = properties();
+	const classes = theme.classes(css);
+	const { messages } = i18n.localize(bundle);
+
+	if (initialPage !== undefined && initialPage !== icache.get('initialPage')) {
+		icache.set('initialPage', initialPage);
+		icache.set('currentPage', initialPage);
+	}
+
+	if (initialPageSize !== undefined && initialPageSize !== icache.get('initialPageSize')) {
+		icache.set('initialPageSize', initialPageSize);
+		icache.set('pageSize', initialPageSize);
+	}
+
+	const page = icache.getOrSet('currentPage', 1);
+	const pageSize = icache.getOrSet('pageSize', 10);
+
+	const pageLink = function(
+		page: number,
+		content?: WNode | WNode[],
+		extraClasses: string = classes.numberedLink,
+		key?: string
+	): WNode {
+		return (
+			<button
+				type="button"
+				key={key || `numberedLink-${page}`}
+				classes={[extraClasses, classes.link]}
+				onclick={(e) => {
+					e.stopPropagation();
+					icache.set('currentPage', page);
+					onPageChange && onPageChange(page);
+				}}
+			>
+				{content || page.toString()}
+			</button>
+		);
+	};
+
+	const leading: WNode[] = [];
+	const leadingCount = Math.min(siblingCount, page - 1);
+	for (let i = 1; i <= leadingCount; i++) {
+		leading.unshift(pageLink(page - i));
+	}
+
+	const trailing: WNode[] = [];
+	const trailingCount = Math.min(siblingCount, total - page);
+	for (let i = 1; i <= trailingCount; i++) {
+		trailing.push(pageLink(page + i));
+	}
+
+	return (
+		total > 1 && (
+			<div key="root" classes={[theme.variant(), classes.root]}>
+				{leadingCount &&
+					pageLink(
+						page - 1,
+						[
+							<div classes={classes.icon}>
+								<Icon type="leftIcon" />
+							</div>,
+							<div classes={classes.label}>{messages.previous}</div>
+						],
+						classes.prev,
+						'prev'
+					)}
+				{...leading}
+				<div classes={classes.currentPage}>{page.toString()}</div>
+				{...trailing}
+				{trailingCount &&
+					pageLink(
+						page + 1,
+						[
+							<div classes={classes.icon}>
+								<Icon type="rightIcon" />
+							</div>,
+							<div classes={classes.label}>{messages.next}</div>
+						],
+						classes.next,
+						'next'
+					)}
+				{pageSizes && pageSizes.length > 0 && (
+					<div classes={classes.selectWrapper}>
+						<Select
+							key="page-size-select"
+							transform={defaultTransform}
+							initialValue={pageSize.toString()}
+							resource={{
+								resource: () => createResource(memoryTemplate),
+								data: pageSizes.map((ps) => ({ value: ps.toString() }))
```

Comment:
this data array should be calculated only when `pageSizes` changes and stored in an `icache` otherwise this will be a new array ever render and cause the resource to be re-populated every time.